src/bentoml/server.py

```diff
@@ -115,75 +119,111 @@ def __init__(
         self.args = args
         self.host = "127.0.0.1" if host == "0.0.0.0" else host
         self.port = port
-        self.start = self._create_startmanager()
-
-    def _create_startmanager(server):  # type: ignore # not calling self self
-        class _StartManager:
-            def __init__(
-                self, blocking: bool = False, env: dict[str, str] | None = None
-            ):
-                """Start the server programmatically.
-
-                To get the client, use the context manager.
-
-                .. note::
-
-                   ``blocking=True`` and using ``start()`` as a context manager is mutually exclusive.
-
-                Args:
-                    blocking: If True, the server will block until it is stopped.
-                    env: A dictionary of environment variables to pass to the server. Default to ``None``.
-                """
-                logger.info(f"starting server with arguments: {server.args}")
 
-                server.process = subprocess.Popen(
-                    server.args,
-                    stdout=subprocess.PIPE,
-                    stderr=subprocess.PIPE,
-                    stdin=subprocess.PIPE,
+    def start(
+        self,
+        blocking: bool = False,
+        env: dict[str, str] | None = None,
+        stdin: _FILE = None,
+        stdout: _FILE = None,
+        stderr: _FILE = None,
+    ):
+        """Start the server programmatically.
+
+        To get the client, use the context manager.
+
+        .. note::
+           ``blocking=True`` and using ``start()`` as a context manager is mutually exclusive.
+
+        Args:
+            blocking: If True, the server will block until it is stopped.
+            env: A dictionary of environment variables to pass to the server. Default to ``None``.
+            stdin: The stdin to pass to the server. Default to ``None``.
+            stdout: The stdout to pass to the server. Default to ``None``.
+            stderr: The stderr to pass to the server. Default to ``None``.
+        """
+
+        class _Manager:
+            def __init__(__inner_self):
+                logger.debug(f"Starting server with arguments: {self.args}")
+                self.process = subprocess.Popen(
+                    self.args,
+                    stdout=None if blocking else (stdout or subprocess.PIPE),
```

Comment:
Or maybe can factor out the second bit.

```suggestion
                    stdout=stdout if stdout is not None else (None if blocking else subprocess.PIPE),
```